File: leetcode/974-subarray-sums-divisible-by-k/Solution.java
class Solution {
    public int subarraysDivByK(int[] A, int K) {
        // map from cumulative sum's modulo to its appear times
        int[] count = new int[K];
        count[0] = 1;
        int ans = 0, sum = 0;
        for (int x : A) {
            sum += x;
            // Redundant + K because in Java mod function 
            // may return negative value
            int idx = (sum % K + K) % K;
            ans += count[idx]++;
        }
        return ans;
    }
}

File: leetcode/974-subarray-sums-divisible-by-k/Solution.py
class Solution:
    def subarraysDivByK(self, A: List[int], K: int) -> int:
        sums = [0]*(len(A)+1)
        for i in range(0, len(A)):
            sums[i+1] = (sums[i] + A[i]) % K
            
        count = collections.Counter(sums)
        return sum(v*(v-1)//2 for v in count.values())
